### DISCUSSION TITLE
RocksDB not working with default config for SMTP only

### DISCUSSION THREAD

**Discussion by 11notes** [stalwartlabs/smtp-server#42] [Q&A] [answered]
RocksDB not working?

```
2024-02-16T07:20:33.226382Z ERROR utils: Invalid configuration: Unable to find data store 'rocksdb' defined in key '[storage.data](http://storage.data/)'
Invalid configuration: Unable to find data store 'rocksdb' defined in key '[storage.data](http://storage.data/)'
```

config used: https://gist.github.com/11notes/6e63a2aab0d65f541f53826769433981

**Comment by mdecimus** [comment]
The `smtp-server` binary does not come with RocksDB support built-in, you can recompile it with RocksDB support by modifying Cargo.toml:

```toml
store = { path = "main/crates/store", features = ["redis", "postgres", "mysql", "sqlite", "rocksdb"] }
```

Or you can use the `mail-server` binary instead.

**Reply by 11notes** [reply]
Thanks, I added rocksdb, and it compiles alongside smtp, but I get still the same error.

```
2024-02-16T22:10:33.439088Z  INFO utils: Starting Stalwart SMTP Server v0.6.0...
2024-02-16T22:10:33.441541Z DEBUG store::config: Unknown directory type: "rocksdb"
2024-02-16T22:10:33.441641Z ERROR utils: Invalid configuration: Unable to find data store 'rocksdb' defined in key 'storage.data'
Invalid configuration: Unable to find data store 'rocksdb' defined in key 'storage.data'
```

https://github.com/11notes/docker-stalwart-smtp/blob/296a228fcad04413a14adb26bdc42792aaf186fd/amd64.dockerfile#L27

If copied the config of the mail-server 1:1 and same result:
```
Invalid configuration: Unable to find data store 'rocksdb' defined in key 'storage.data'
```

If I use postgres it all works, rocksdb not.

**Reply by mdecimus** [reply]
> 2024-02-16T22:10:33.441541Z DEBUG store::config: Unknown directory type: "rocksdb"

The problem here is that you are using an invalid directory type, it has to be of type `internal` and then reference the RocksDb store as follows:

```toml
[directory."mydirectory"]
type = "internal"
store = "rocksdb"

[store."rocksdb"]
type = "rocksdb"
...
```

**Comment by 11notes** [comment]
Okay, I use postgres now, that works, I don't know why rocksdb does not work, but for my use case it doesn't matter, what matters is Redis, and that also doesn't work:

LOG
```sh
2024-02-17T01:04:21.609143Z  WARN smtp::core::eval: Failed to get key. context="eval_if" event="error" property="session.rcpt.relay" error=InternalError("Redis error: NOAUTH: Authentication required.")
```

REDIS
```sh
acl setuser stalwart on >stalwart allkeys +@all
```

redis.toml
```toml
[store."redis"]
type = "redis"
url = "redis://localhost"
username = "stalwart"
password = "stalwart"
timeout = "10s"
retries = 3
max-retry-wait = "1s"
min-retry-wait = "500ms"
```

I would also like to use rediss (TLS) but if I do that with a self signed certificate there is no configuration flag to disable SSL verification for Redis. Any ideas? It would exactly be better if you could utilize the full redis URl with username, password, DB selector, and so on, like described here: https://github.com/lettuce-io/lettuce-core/wiki/Redis-URI-and-connection-details

**Reply by mdecimus** [reply, marked as answer]
The Redis library that Stalwart uses is configured differently depending on whether you're using Redis or Redis cluster. I'll have this better documented, but Redis is configured entirely from the URL using the following syntax:

```redis://[<username>][:<password>@]<hostname>[:port][/<db>]```

And to use `rediss` with a self-signed certificate:

```rediss://[<username>][:<password>@]<hostname>[:port][/<db>]#insecure```

**Edit**: The Redis documentation has been [updated](https://stalw.art/docs/storage/backends/redis/).

**Reply by 11notes** [reply]
Thanks, that worked like the charm. During my tests I’ve seen that many of the options available in the default configuration are not documented in the documentation. Would be nice if this would be one day on par.

**Reply by mdecimus** [reply]
Which settings have you been unable to find in the documentation?
